0017-sampler-interface.md

```diff
@@ -0,0 +1,282 @@
+ * _Has no sample weights._ The current sampler attaches real weights summing to one to all samples. In effect, it returns an estimate of a (quasi)-distribution instead of samples from the distribution. This proposal eliminates this feature and returns only samples, and without weights. The rationale is to make the `Sampler` into more of a streamlined circuit executor, paring down all frills.
+
+
+## User Benefit <a name="user-benefit"></a>
+
+Users will be able to transition from `backend.run` to `sampler.run` without any compromises in execution efficiency. They will be able to run dynamic circuits without having to specify special flags, or invoke a special program. They will get their outputs in a format where they don't need to manually separate bitstrings from different registers.
+
+As with the [(paired Estimator inteface RFC)[https://github.com/Qiskit/RFCs/pull/51]](https://github.com/Qiskit/RFCs/pull/51), implementations will more easily be able to enable features like late stage parameter value binding.
+
+## Design Proposal <a name="design-proposal"></a>
+
+We will lead the design proposal with an example.
+
+```python
+# Our first circuit has many Parameters, but elects to do all measurement at 
+# the end. Note that measure_all() chooses the register name "meas" for you.
+circuit1 = QuantumCircuit(9)
+...
+<use 2039 Parameters in this circuit>
+...
+circuit1.measure_all()
+
+# Specify 128 different parameter value sets for circuit1, in a 32x4 shape.
+parameter_values1 = np.random.uniform(size=(32, 4, 2039))
+
+# Our second circuit happens to have no parameters, but three output registers 
+# named alpha, beta, and gamma of various sizes
+alpha = ClassicalRegister(127, "alpha")
+beta = ClassicalRegister(10, "beta")
+gamma = ClassicalRegister(15, "gamma")
+circuit2 = QuantumCircuit(QuantumRegister(127), alpha, beta, gamma)
+...
+<insert gates, measure into the registers, apply dynamic circuit features, etc.>
+...
+
+# invoke a sampler
+sampler = Sampler(..., shots=1024)
+job = sampler.run([
+    (circuit1, parameter_values1),
+    circuit2
+])
+
+# wait for the result
+job.result()
+
+# get a bundle of results for every task
+>> PrimitiveResult(
+>>     SamplerTaskResult(
+>>         DataBundle(meas=BitArray(<num_samples=1024, num_bits=127, shape=(32, 4)>)), 
+>>         <metadata>
+>>     ), 
+>>     SamplerTaskResult(
+>>         DataBundle(
+>>             alpha=BitArray(<num_samples=1024, num_bits=127, shape=()>),
+>>             beta=BitArray(<num_samples=1024, num_bits=10, shape=()>),
+>>             gamma=BitArray(<num_samples=1024, num_bits=15, shape=()>)
+>>         ), 
+>>         <metadata>
+>>     ), 
+>>     metadata=<global_metadata>
+>> )
```

Comment:
Two questions about the returned result datastructures:
- Is this also compatible if, in the future, we want to enable streaming of results? do we need to change the datastructures at all?
- similar to Estimator, I think it would be good to add a field here which is the confidence interval, and it gets smaller with more shots.

Reply:
1) I imagine that a streaming interface would look substantially different, and would require new tooling. The emphasis here is on data contiguity for fast local operations. I think it would probably be best to treat that with separate tools.

2) I'm wondering what the confidence interval would be of? The observations are the observations

Reply:
You could store such information in the task result metadata.

Reply:
I will note that the IBM hardware discriminators don't return enough information to provide this information in general. You need to activate the `meas_level=1` interface in `backend.run` to have access to that data today.